File: python/robot-simulator/robot_simulator.py
DIRECTIONS = [(1, 0), (0, 1), (-1, 0), (0, -1)]
EAST, NORTH, WEST, SOUTH = DIRECTIONS

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, commands):
        for cmd in commands:
            match cmd:
                case 'R': self.turn_right()
                case 'L': self.turn_left()
                case 'A': self.advance()

    def turn_right(self):
        self.direction = DIRECTIONS[(DIRECTIONS.index(self.direction) - 1) % 4]

    def turn_left(self):
        self.direction = DIRECTIONS[(DIRECTIONS.index(self.direction) + 1) % 4]

    def advance(self):
        x, y = self.coordinates
        px, py = self.direction
        self.coordinates = (x + px, y + py)
        

File: typescript/robot-simulator/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

const DIRECTION: Record<Direction, Coordinates> = {
  north: [0, 1],
  east: [1, 0],
  south: [0, -1],
  west: [-1, 0],
}

export class Robot {
  pos: Coordinates = [0, 0]
  dir: Direction = 'north'
  
  get bearing(): Direction {
    return this.dir
  }

  get coordinates(): Coordinates {
    return this.pos
  }

  place({x, y, direction = 'north'}: { x: number; y: number; direction: string }) {
    if (!['north', 'east', 'south', 'west'].includes(direction)) {
      throw new InvalidInputError('')
    }
    
    this.pos = [x, y]
    this.dir = direction as Direction
  }

  evaluate(instructions: string) {
    const dirs = Object.keys(DIRECTION)
    for (const cmd of instructions) {
      if (cmd == 'A') {
        const [mx, my] = DIRECTION[this.bearing]
        const [x, y] = this.pos
        this.pos = [x + mx, y + my]
        continue
      }

      const rotate = cmd == 'R' ? 1 : -1
      this.dir = dirs[(dirs.indexOf(this.dir) + rotate + 4) % 4] as Direction
    }
  }
}
